File: CpuA32/disassembler.cc
// (c) Robert Muth - see LICENSE for more info

#include "CpuA32/disassembler.h"
#include "Util/assert.h"
#include "Util/parse.h"

#include <stdio.h>
#include <cstring>
#include <string_view>

namespace cwerg {
namespace a32 {

const char* kArmRegNames[] = {                       //
    "r0", "r1", "r2", "r3", "r4", "r5", "r6", "r7",  //
    "r8", "r9", "sl", "fp", "ip", "sp", "lr", "pc"};

const constexpr std::string_view kNumZero("#0");

// TODO: stop using printf and realted cruft from <cstring>

char* strappend(char* dst, const char* src) {
  strcpy(dst, src);
  return dst + strlen(dst);
}

char* strappenddec(char* dst, int64_t n) {
  if (n >= 0) {
    ToDecString(n, dst);
  } else {
    *dst++ = '-';
    ToDecString(-n, dst);
  }
  return dst + strlen(dst);
}

char* strappendhex(char* dst, int64_t n) {
  ToHexString(n, dst);
  return dst + strlen(dst);
}

char* strappend(char* dst, std::string_view src) {
  memcpy(dst, src.data(), src.size());
  dst[src.size()] = 0;
  return dst + src.size();
}

int RenderMultipleVFP(const char* reg_prefix,
                      int32_t start_reg,
                      int32_t count,
                      char* buffer) {
  ASSERT(count > 0, "");
  if (count == 1) {
    buffer = strappend(buffer, "{");
    buffer = strappend(buffer, reg_prefix);
    buffer = strappenddec(buffer, start_reg);
    buffer = strappend(buffer, "}");
  } else {
    sprintf(buffer, "{%s%d-%s%d}", reg_prefix, start_reg, reg_prefix,
            start_reg + count - 1);
  }
  return 1;
}

char* RenderOperandSystematic(char* buffer, int32_t x, OK ok) {
  switch (ok) {
    case OK::PRED_28_31:
      return strappend(buffer, EnumToString(PRED(x)));
    case OK::REG_0_3:
    case OK::REG_12_15:
    case OK::REG_16_19:
    case OK::REG_8_11:
    case OK::REG_PAIR_12_15:
      ASSERT(x <= 15, "REG out of range " << x);
      return strappend(buffer, kArmRegNames[x]);
    case OK::DREG_0_3_5:
    case OK::DREG_12_15_22:
    case OK::DREG_16_19_7:
      ASSERT(x <= 15, " DREG out of range " << x);
      *buffer++ = 'd';
      return strappenddec(buffer, x);
    case OK::SREG_0_3_5:
    case OK::SREG_12_15_22:
    case OK::SREG_16_19_7:
      ASSERT(x <= 31, "SREG out of range " << x);
      *buffer++ = 's';
      return strappenddec(buffer, x);
    case OK::IMM_0_11:
    case OK::IMM_0_11_16_19:
    case OK::IMM_0_3_8_11:
    case OK::IMM_0_7_8_11:
    case OK::IMM_0_7_TIMES_4:
    case OK::IMM_10_11_TIMES_8:
    case OK::IMM_7_11:
    case OK::IMM_0_23:
    case OK::IMM_FLT_ZERO:
      sprintf(buffer, "%u", (uint32_t)x);
      return buffer + strlen(buffer);
    case OK::SIMM_0_23:
      return strappenddec(buffer, x);
    case OK::REG_RANGE_0_7:
    case OK::REG_RANGE_1_7:
      buffer = strappend(buffer, "regrange:");
      return strappenddec(buffer, x);
    case OK::REGLIST_0_15:
      sprintf(buffer, "reglist:0x%04x", x);
      return buffer + strlen(buffer);
    case OK::SHIFT_MODE_5_6:
      return  strappend(buffer, EnumToString(SHIFT(x)));
    case OK::Invalid:
    default:
      ASSERT(false, "unhandled OK: " << (unsigned)ok);
      return buffer;
  }
}

// render a single operand, e.g. and address like  `[r3, #-116]`
// Used to sanity check against objdump output
char* RenderOperandStd(char* buffer, const Opcode& opcode, int32_t x, OK ok) {
  switch (ok) {
    case OK::REG_0_3:
      if (opcode.classes & OPC_FLAG::ADDR_DEC) *buffer++ = '-';
      // FALL-THROUGH
    case OK::REG_12_15:
    case OK::REG_16_19:
    case OK::REG_8_11:
    case OK::REG_PAIR_12_15:
    case OK::DREG_0_3_5:
    case OK::DREG_12_15_22:
    case OK::DREG_16_19_7:
    case OK::SREG_0_3_5:
    case OK::SREG_12_15_22:
    case OK::SREG_16_19_7:
    //
    case OK::PRED_28_31:
    case OK::SHIFT_MODE_5_6:
      return RenderOperandSystematic(buffer, x, ok);
    case OK::IMM_0_7_TIMES_4:
    case OK::IMM_0_11:
    case OK::IMM_0_3_8_11:
      *buffer++ = '#';
      if (opcode.classes & OPC_FLAG::ADDR_DEC) {
        *buffer++ = '-';
      }
      return strappenddec(buffer, x);
    case OK::IMM_0_11_16_19:
    case OK::IMM_0_7_8_11:
    case OK::IMM_10_11_TIMES_8:
    case OK::IMM_7_11:
    case OK::IMM_0_23:
    case OK::IMM_FLT_ZERO:
    case OK::SIMM_0_23:
    case OK::REG_RANGE_0_7:
    case OK::REG_RANGE_1_7:
      *buffer++ = '#';
      return strappenddec(buffer, x);
    case OK::REGLIST_0_15: {
      const char* sep = "";
      buffer = strappend(buffer, "{");
      for (int i = 0; i < 16; ++i) {
        if (x & (1 << i)) {
          buffer = strappend(buffer, sep);
          buffer = strappend(buffer, kArmRegNames[i]);
          sep = ", ";
        }
      }
      buffer = strappend(buffer, "}");
    }
      return buffer;
    case OK::Invalid:
    default:
      ASSERT(false, "Invalid");
      return buffer;
  }
}


// Render the given instruction using a systematic notation into `buffer`
void RenderInsSystematic(const Ins& ins, char buffer[512]) {
  buffer = strappend(buffer, ins.opcode->enum_name);
  buffer = strappend(buffer, " ");

  const char* sep = "";
  for (unsigned i = 0; i < ins.opcode->num_fields; ++i) {
    buffer = strappend(buffer, sep);
    if (ins.reloc_kind != elf::RELOC_TYPE_ARM::NONE && i == ins.reloc_pos) {
      switch (ins.reloc_kind) {
        case elf::RELOC_TYPE_ARM::JUMP24:
          ASSERT(ins.is_local_sym, "");
          buffer = strappend(buffer, "expr:jump24:");
          buffer = strappend(buffer, ins.reloc_symbol);
          break;
        case elf::RELOC_TYPE_ARM::CALL:
          buffer = strappend(buffer, "expr:call:");
          buffer = strappend(buffer, ins.reloc_symbol);
          break;
        case elf::RELOC_TYPE_ARM::MOVT_ABS:
          buffer = strappend(buffer, ins.is_local_sym ? "expr:loc_movt_abs:"
                                                      : "expr:movt_abs:");
          buffer = strappend(buffer, ins.reloc_symbol);
          break;
        case elf::RELOC_TYPE_ARM::MOVW_ABS_NC:
          buffer = strappend(buffer, ins.is_local_sym ? "expr:loc_movw_abs_nc:"
                                                      : "expr:movw_abs_nc:");
          buffer = strappend(buffer, ins.reloc_symbol);

          break;
        default:
          ASSERT(false, "");
      }
      if (ins.operands[i] != 0) {
        *buffer++ = ':';
        buffer = strappenddec(buffer, ins.operands[i]);
      }
    } else {
      buffer = RenderOperandSystematic(buffer, ins.operands[i],
                                       ins.opcode->fields[i]);
    }
    sep = " ";
  }
}

}  // namespace a32
}  // namespace cwerg

File: CpuA32/disassembler.py
#!/usr/bin/python3

"""
Convert ARM32 instruction into human readable form suitable to
be processed by an assembler.
"""

import CpuA32.opcode_tab as arm
from Elf import enum_tab

from typing import List


def _Merge(*name_lists):
    out = {}
    for nl in name_lists:
        for n, name in enumerate(nl):
            x = out.get(name)
            if x is None:
                out[name] = n
            else:
                assert x == n, f"conflict {n} vs {x} for {name}"
    return out


PRED_NAMES_MAP = _Merge([p.name for p in arm.PRED])

REG_NAMES_MAP = _Merge([p.name for p in arm.REG],
                       [f"r{i}" for i in range(16)],
                       [p.name for p in arm.DREG],
                       [p.name for p in arm.SREG])

SHIFT_NAMES_MAP = _Merge([p.name for p in arm.SHIFT])


# render a single operand, e.g. and address like  `[r3, #-116]`
def RenderOperandStd(opcode: arm.Opcode, operand, ok) -> str:
    if ok is arm.OK.REG_LINK:
        return "lr"
    elif ok in arm.FIELDS_SREG:
        return arm.SREG(operand).name
    elif ok in arm.FIELDS_DREG:
        return arm.DREG(operand).name
    elif ok is arm.OK.REG_0_3 and arm.OPC_FLAG.ADDR_DEC in opcode.classes:
        return "-" + arm.REG(operand).name
    elif ok in arm.FIELDS_REG:
        return arm.REG(operand).name
    elif ok is arm.OK.SHIFT_MODE_5_6:
        return arm.SHIFT(operand).name
    elif ok in {arm.OK.IMM_0_7_TIMES_4, arm.OK.IMM_0_11, arm.OK.IMM_0_3_8_11}:
        if arm.OPC_FLAG.ADDR_DEC in opcode.classes:
            return f"#-{operand}"
        else:
            return f"#{operand}"
    elif ok in arm.FIELDS_IMM:
        return f"#{operand}"
    elif ok is arm.OK.PRED_28_31:
        return arm.PRED(operand).name
    elif ok is arm.OK.REGLIST_0_15:
        reg_mask = operand
        regs = [arm.REG(x).name for x in range(16) if reg_mask & (1 << x)]
        expr = "{%s}" % (",".join(regs))
        return expr
    elif ok is arm.OK.REG_RANGE_0_7 or ok is arm.OK.REG_RANGE_1_7:
        return f"#{operand}"
    else:
        assert False, f"OK:{ok} OP:{operand}"


def _EmitReloc(ins: arm.Ins, pos: int) -> str:
    if ins.reloc_kind == enum_tab.RELOC_TYPE_ARM.JUMP24:
        assert ins.is_local_sym, f"expected local symbol"
        return f"expr:jump24:{ins.reloc_symbol}"
    elif ins.reloc_kind == enum_tab.RELOC_TYPE_ARM.MOVT_ABS:
        loc = "loc_" if ins.is_local_sym else ""
        offset = "" if ins.operands[pos] == 0 else f":{ins.operands[pos]}"
        return f"expr:{loc}movt_abs:{ins.reloc_symbol}{offset}"
    elif ins.reloc_kind == enum_tab.RELOC_TYPE_ARM.MOVW_ABS_NC:
        loc = "loc_" if ins.is_local_sym else ""
        offset = "" if ins.operands[pos] == 0 else f":{ins.operands[pos]}"
        return f"expr:{loc}movw_abs_nc:{ins.reloc_symbol}{offset}"
    elif ins.reloc_kind == enum_tab.RELOC_TYPE_ARM.CALL:
        return f"expr:call:{ins.reloc_symbol}"
    else:
        assert False


def _SymbolizeOperand(field: arm.OK, value: int) -> str:
    if field == arm.OK.PRED_28_31:
        return arm.PRED(value).name
    elif field in arm.FIELDS_REG:
        return arm.REG(value).name
    elif field in arm.FIELDS_DREG:
        return arm.DREG(value).name
    elif field in arm.FIELDS_SREG:
        return arm.SREG(value).name
    elif field in arm.FIELDS_IMM:
        return f"{value}"
    elif field in arm.FIELDS_SHIFT:
        return arm.SHIFT(value).name
    elif field is arm.OK.REG_RANGE_0_7 or field is arm.OK.REG_RANGE_1_7:
        return f"regrange:{value}"
    elif field is arm.OK.REGLIST_0_15:
        return f"reglist:0x{value:04x}"
    else:
        assert False, f"unsupported field {field}"
        return ""


def SymbolizeOperands(ins: arm.Ins) -> List[str]:
    out = []
    for pos, (field, value) in enumerate(zip(ins.opcode.fields, ins.operands)):
        if field in arm.FIELDS_IMM and ins.reloc_kind != 0 and ins.reloc_pos == pos:
            out.append(_EmitReloc(ins, pos))
        else:
            out.append(_SymbolizeOperand(field, value))
    return out


def RenderInstructionSystematic(ins: arm.Ins) -> str:
    return f"{ins.opcode.NameForEnum()} {' '.join(SymbolizeOperands(ins))}"


def UnsymbolizeOperand(o: str, ok: arm.OK) -> int:
    if ok is arm.OK.SHIFT_MODE_5_6:
        return SHIFT_NAMES_MAP[o]
    elif ok is arm.OK.PRED_28_31:
        return PRED_NAMES_MAP[o]
    x = REG_NAMES_MAP.get(o)
    if x is not None:
        return x
    elif ":" in o:
        tag, val = o.split(":")
        return int(val, 0)
    else:
        # must be a number
        return int(o, 0)


_RELOC_KIND_MAP = {
    "abs32": enum_tab.RELOC_TYPE_ARM.ABS32,
    "jump24": enum_tab.RELOC_TYPE_ARM.JUMP24,
    "call": enum_tab.RELOC_TYPE_ARM.CALL,
    "movw_abs_nc": enum_tab.RELOC_TYPE_ARM.MOVW_ABS_NC,
    "movt_abs": enum_tab.RELOC_TYPE_ARM.MOVT_ABS,
}


def InsParse(mnemonic, token: List[str]) -> arm.Ins:
    """Takes textual form of an A32 instruction (in systematic notation) and parses it into a Ins

     * Supports relocatable expressions
     * Adds missing "al" predicate if necessary
     Example input:
     "add_regimm", ["r4", "r4", "r0", "lsl", "0"]
     """
    opcode: arm.Opcode = arm.Opcode.name_to_opcode[mnemonic]
    if opcode.HasPred() and len(token) == len(opcode.fields) - 1:
        token = ["al"] + token
    ins = arm.Ins(opcode)
    for pos, (t, ok) in enumerate(zip(token, opcode.fields)):
        if t.startswith("expr:"):
            # expr strings have the form expr:<rel-kind>:<symbol>:<addend>, e.g.:
            #   expr:movw_abs_nc:string_pointers:5
            #   expr:call:putchar
            rel_token = t.split(":")
            if len(rel_token) == 3:
                rel_token.append("0")
            if rel_token[1] == "jump24":
                ins.is_local_sym = True
            if rel_token[1].startswith("loc_"):
                ins.is_local_sym = True
                rel_token[1] = rel_token[1][4:]
            ins.reloc_kind = _RELOC_KIND_MAP[rel_token[1]]
            ins.reloc_pos = pos
            ins.reloc_symbol = rel_token[2]
            ins.operands.append(int(rel_token[3], 0))
        else:
            ins.operands.append(UnsymbolizeOperand(t, ok))
    return ins


if __name__ == "__main__":
    import sys


    def disass(data):
        ins = arm.Disassemble(data)
        if ins.opcode is None:
            print(f"could not disassemble {data:x}")
            return
        std_name = ins.opcode.name
        std_ops = [RenderOperandStd(ins.opcode, op, ok)
                   for ok, op in zip(ins.opcode.fields, ins.operands)]

        if std_ops and std_ops[0] == "al":
            std_ops.pop(0)
        print(f"{data:08x} {std_name} {', '.join(std_ops)}")
        print("OPCODE", ins.opcode.name, ins.opcode.variant)
        operands_str = SymbolizeOperands(ins)
        for f, o, o_str in zip(ins.opcode.fields, ins.operands, operands_str):
            print(f"    {f.name:19s} {o_str} ({o})")
        print()
        data2 = arm.Assemble(ins)
        assert data == data2
        operands2 = [UnsymbolizeOperand(o, ok)
                     for o, ok in zip(operands_str, ins.opcode.fields)]
        assert tuple(ins.operands) == tuple(

            operands2), f"{ins.operands} vs {operands2}"


    # opcode.AssembleOperands(operands2)

    for arg_hex_number in sys.argv[1:]:
        disass(int(arg_hex_number, 16))
